Sources/SPMTestSupport/MockPackage.swift

```diff
@@ -20,6 +20,8 @@ public struct MockPackage {
     public let products: [MockProduct]
     public let dependencies: [MockDependency]
     public let versions: [String?]
+    /// Revision identifier for the package contents. A random identifier might be assigned if this is nil.
+    public let revision: String?
```

Comment:
this is interesting. I understand why it was added (to control the hash in the mock use case) and that makes sense, but its not really a package level setting - its more of a version specific setting. i.e. `versions: [String?] = []` may need to become more complex to support it and then MockWorkspace would need to "commit" on every version where the revision/hash was changed. it makes things more complicated but a more accurate model. another, maybe simpler, alternative is for the MockPackage to take an optional `revisionProvider: (version: String) -> String` which would return the hash for the version and then MockWorkspace can again commit on hash changes per version.

Reply:
> but its not really a package level setting - its more of a version specific setting

This is what I started with until I saw we "commit" only once but tag for each version in `MockWorkspace`. I will try to update that logic.

Reply:
Another option could be to make `versions` an array of tuples?

Reply:
since version is used by many call sites, I personally would do a `revisionProvider: (version: String) -> String` which would rarely be used

Reply:
https://github.com/apple/swift-package-manager/pull/3890/commits/f89adab73e83cfe3d1d340a1c790d2998ba9e5e4